specs/BUNDLE_SPEC.md

```diff
@@ -0,0 +1,271 @@
+```
+
+A `404 Not Found` response indicates that the registry does not support the
+referrers API and the referrers tag scheme should be used as a fallback:
+
+```
+GET /v2/foo/manifests/sha256-c000100ff...
+```
+
+A `404` here would indicate that there are no artifacts associated with the
+image.
+
+Assuming there are artifacts present, one of the two above calls will return an
+image index listing the artifacts which have been associated with the specified
+image:
+
+```
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.index.v1+json",
+  "manifests": [
+    {
+      "mediaType": "application/vnd.oci.image.manifest.v1+json",
+      "digest": "sha256:badf00d..",
+      "size": 779,
+      "artifactType": "application/vnd.dev.sigstore.bundle+json;version=0.2"
+    }
+  ]
+}
+```
+
+From this the client can identify any Sigstore bundles by looking at the
+`artifactType` field.
+
+Using the `digest` listed in the image index, the next step is to retrieve the
+manifest for the bundle:
+
+```
+GET /v2/foo/manifests/sha256:badf00d..
+```
+
+```
+{
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "schemaVersion": 2,
+  "artifactType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
+  "config": {
+    "mediaType": "application/vnd.oci.empty.v1+json",
+    "digest": "sha256:44136fa3...",
+    "size": 2
+  },
+  "layers": [
+    {
+      "digest": "sha256:cafed00d...",
+      "mediaType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
+      "size": 4971
+    }
+  ],
+  "subject": {
+    "digest": "sha256:c00010ff...",
+    "mediaType": "application/vnd.oci.image.index.v1+json"
+   }
+}
+```
+
+The final step is to use the `digest` from the first of the `layers` to retrieve
+the bundle blob:
+
+```
+GET /v2/foo/blobs/uploads/?digest=cafed00d...
+```
+
+```
+{
+  "mediaType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
+  "verificationMaterial": {...},
+  "messageSignature": {...}
+}
+```
+
+## Annotations
+
+For any given image, there may be any number of attached attestation bundles.
+When there are multiple Sigstore bundles associated with an image it may be
+difficult to identify which artifact is which in the image index:
+
+```json
+{
+  "mediaType": "application/vnd.oci.image.index.v1+json",
+  "schemaVersion": 2,
+  "manifests": [
+    {
+      "artifactType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
+      "digest": "sha256:facefeed",
+      "mediaType": "application/vnd.oci.image.manifest.v1+json"
+    },
+    {
+      "artifactType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
+      "digest": "sha256:d0d0caca",
+      "mediaType": "application/vnd.oci.image.manifest.v1+json"
+    }
+  ]
+}
+```
+
+To help help disambiguate attestations, clients may add annotations to the items
+in the `manifests` list which indicate what is contained within each bundle:
+
+- `dev.sigstore.cosign.bundle.content` - Must be one "message-signature" or
```

Comment:
You mention at the beginning this being used outside of Cosign. If so, should we have an annotation without `cosign`, maybe just `dev.sigstore.bundle.content` or `dev.sigstore.container.bundle.content` or something?

Reply:
I was sorta following the lead of the [signature spec](https://github.com/sigstore/cosign/blob/main/specs/SIGNATURE_SPEC.md) which uses things like `dev.sigstore.cosign/certificate` and `dev.sigstore.cosign/chain` . . . but would be happy to ditch the `cosign` bit to make it clear that creating/consuming these attestations is not dependent on cosign in any way.

Reply:
Yea, I assumed this was for consistency. Let's remove it, it'll make it easier to adopt with other language implementations.

Reply:
Removed the `cosign` segment from the annotation keys